bitcoin-core-dev-tech/2023-09/p2p-working-session.md

```diff
@@ -0,0 +1,204 @@
+- Only broadcast to Tor peers if possible. Send to someone random and disconnect. If I only tell Tor peers and a safety net to tell if it is propagated and if not broadcast.
+- Can see the timing and where txs originate from.
+- The trickling mechanism adds randomness and doesn't obscure things from an attacker that sees everything
+- Inspired by [Linking Lion](https://b10c.me/observations/05-inbound-connection-flooder-down/)
+- Problem is the change and where to put that.
+- If we send it to a few peers over Tor and some of them are spies, they have a privileged position to see. If they come to our mempool and ask and they can confirm if it is yours. Then if some asks and then we pretend we don't have it.
+  - you are changing the behavior of what we allow in the mempool.
+  - If some sends an INV or sends GETDATA - parent and child issues also complicate things. Convinced I can hide it.
+- Now I'm leaning the other way, which is not to put it in the mempool. But where? Need to put it on disk. In the mempool there are unbroadcasted txs. The code is already there. If it doesn't come back, you need to put it somewhere.
+- There are two cases to think about. It comes from your wallet or it's from RPC. What happens if the RPC comes from the wallet?
+- SendoutTx would put in your mempool. Subject to trickling mechanism will go in the ToBeINVed.
+  - What about announcing to some of your peers, wait to see it echoed back and then send to all.
+  - How bitcoin core worked until 0.5
+  - It's not trivial I think. Need to act differently for that Txs than Txs coming from elsewhere which may also be a tell. And also need a more aggressive broadcast mechanism until it is echoed back.
+  - we don't want to send it to everyone. Then we send through Tor, then you could skip all the rest and then you destroy the assumption of the attacker.
+  - if you fluff it, then someone can assume that you are the source wallet. Even if none of the behavior changes, it destroys the assumption reliability. It lowers the confidence level.
+  - chainalysis can't see where it originates
+- if you have a time-sensitive protocol, that might be an issue
+- after 20 min try the same
+  - 20 min seems enormous. Should only be a few mins. An attacker watching all the connections. If you only send it to 1 then they can't do the timing connection. It would be nice as a stretch goal, to obscure that the mechanism is being used.
+  - In addition to Tor peers, send to clearnet connections through Tor. In I2P, no source connections.
+- can you coin flip this, by coinflip
+- if you do it twice then they can tell it is yours
+- what do you think about sending out once immediately and after a short delay (seconds) do it again.
+- This happens now because we have only 2 threads for the connection manager. So it is going to take some time. Should I broadcast it to 5 unconditionally or 1 and as soon as
```

Comment:
```suggestion
- This happens now because we have only 2 threads for the connection manager. So it is going to take some time. Should I broadcast it to 5 unconditionally or 1, then another one and then another one and stop as soon as the transaction is echoed back.
```